# Deriving New Variables

## This Script Includes the Following Points:

### 1. Create 'Price Label' Column

### 2. Create 'Busiest Day' Column

### 3. (Step 2 & 3) Create 'Busiest Days' Column

### 4. (Step 4 & 5) Create 'Busiest Period of Day' Column

In [3]:
#Import Libraries
import pandas as pd
import numpy as np
import os

In [4]:
#Import Data
path = r'C:\Users\Marlo\OneDrive - South Washington County Schools\Home\job support\Coursework\Achievement 4\Instacart Basket Analysis'

In [5]:
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [7]:
#Create Subset
df = df_ords_prods_merge[:1000000]

### 1. Create 'Price Label' Column

In [8]:
#Define a function using user-defined functions

def price_label(row):
    
    if row['prices'] <= 5:
        return 'Low-range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15):
        return 'Mid-range product'
    elif row['prices'] > 15:
        return 'High Range'
    else: return 'Not enough data'

In [11]:
#Check Output
df.shape

(1000000, 14)

In [10]:
df.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  add_to_cart_order  reordered _merge  
0                     3.0                  5          0   both  
1                    20.0                  1          1   both  
2                     6.0                 20          0   both  
3                     NaN                 10          0   both  
4                     9.0                 11          1   both

In [12]:
#Apply the function

df['price_range'] = df.apply(price_label, axis = 1)

C:\Users\Marlo\AppData\Local\Temp\ipykernel_10888\2636238781.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis = 1)


In [13]:
#Check Output
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    652638
Low-range product    338018
High Range             9344
Name: count, dtype: int64

In [14]:
df['prices'].max()

24.5

In [15]:
#Define function using loc() function
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\Marlo\AppData\Local\Temp\ipykernel_10888\2103938361.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'
C:\Users\Marlo\AppData\Local\Temp\ipykernel_10888\2103938361.py:2: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'High-range product' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [16]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [17]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [18]:
#Check Outputs
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     652638
Low-range product     338018
High-range product      9344
Name: count, dtype: int64

In [19]:
#Define function for df_ords_prods_merged
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] = "High-range product"

C:\Users\Marlo\AppData\Local\Temp\ipykernel_10888\1495077028.py:2: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'High-range product' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] = "High-range product"


In [20]:
df_ords_prods_merge.loc[(df_ords_prods_merge['prices'] <= 15) & (df_ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [21]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] < 5, 'price_range_loc'] = 'Low-range product'

In [22]:
#Check Outputs
df_ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21861558
Low-range product      9904846
High-range product      412551
NaN                     225904
Name: count, dtype: int64

#### I'm not sure why there are 225,904 with missing data. It doesn't match the exercise.

### 2. Create 'Busiest Day' Column

In [25]:
#Check Outputs
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [30]:
#Assign Busiest Days and Least Busiest Days
result = []

for value in df_ords_prods_merge["orders_day_of_week"]:
    if value == 0:
        result.append("Busiest Day")
    elif value == 4:
        result.append("Least busy")
    else:
        result.append("Regularly busy")

In [31]:
result

['Regularly busy',
 'Regularly busy',
 'Busiest Day',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest Day',
 'Busiest Day',
 'Busiest Day',
 'Busiest Day',
 'Busiest Day',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest Day',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest Day',
 'Regularly busy',
 'Busiest Day',
 'Least busy',
 'Busiest Day',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Bus

In [33]:
#Create "Busiest Day" Column
df_ords_prods_merge['busiest_day'] = result

In [34]:
#Check Outputs
df_ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest Day        6204182
Least busy         3783802
Name: count, dtype: int64

### 3. (Step 2) Create 'Busiest Days' Column

In [40]:
#Assign Busiest Days and Least Busiest Days
result = []

for value in df_ords_prods_merge["orders_day_of_week"]:
    if value == 0 or value == 1:
        result.append("Busiest Days")
    elif value == 4 or value == 3:
        result.append("Least busy")
    else:
        result.append("Regularly busy")

In [41]:
result

['Regularly busy',
 'Regularly busy',
 'Busiest Days',
 'Least busy',
 'Least busy',
 'Busiest Days',
 'Regularly busy',
 'Least busy',
 'Busiest Days',
 'Busiest Days',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest Days',
 'Busiest Days',
 'Regularly busy',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Least busy',
 'Regularly busy',
 'Busiest Days',
 'Busiest Days',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Least busy',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest Days',
 'Busiest Days',
 'Busiest Days',
 'Least busy',
 'Regularly busy',
 'Busiest Days',
 'Regularly busy',
 'Busiest Days',
 'Busiest Days',
 'Regularly busy',
 '

In [42]:
#Create "Busiest Days" Column
df_ords_prods_merge['busiest_days'] = result

In [43]:
#Check Outputs
df_ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy    12916111
Busiest Days      11864412
Least busy         7624336
Name: count, dtype: int64

Busiest days should be the sum of days "0" and "1". It is.
Least busy days should be the sum of days "4" and "3". It is. 
Regularly busy days should be the sum of days "2", "5", and "6". It is. The loop was operated correctly.

### 4. (Step 4 & 5) Create 'Busiest Period of Day' Column

In [44]:
#Check Outputs
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  add_to_cart_order  reordered _merge  \
0                     3.0                  5          0   both   
1                    20.0                  1          1   both   
2                     6.0                 20          0   both   
3                     NaN                 10          0   both   
4                     9.0                 11          1   both   

     price_range_loc     busiest_day    busiest_days  
0  Mid-range product  Regularly busy  Regularly busy  
1  Mid-range product  Regularly busy  Regularly busy  
2  Mid-range product     Busiest Day    Busiest Days  
3  Mid-range product  Regularly busy      Least busy  
4  Mid-range product      Least busy      Least busy

In [46]:
df_ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

There are 24 hours in the day. If the hours are split equally into the categories "Most Orders", "Average Orders", and "Fewest Orders", the hours would line up as:
"Most Orders" = 10, 11, 14, 15, 13, 12, 16, 9
"Average Orders" = 17, 8, 18, 19, 20, 7, 21, 22
"Fewest Orders" = 23, 6, 0, 1, 5, 2, 4, 3

In [48]:
#Assign Busiest Hours and Least Busiest Houses
result = []

for value in df_ords_prods_merge["order_hour_of_day"]:
    if value in [10, 11, 14, 15, 13, 12, 16, 9]:
        result.append("Most Orders")
    elif value in [17, 8, 18, 19, 20, 7, 21, 22]:
        result.append("Average Orders")
    elif value in [23, 6, 0, 1, 5, 2, 4, 3]:
        result.append("Fewest Orders")
    else:
        result.append("Not Enough Data")

In [49]:
result

['Most Orders',
 'Average Orders',
 'Average Orders',
 'Most Orders',
 'Average Orders',
 'Average Orders',
 'Most Orders',
 'Most Orders',
 'Average Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Average Orders',
 'Average Orders',
 'Fewest Orders',
 'Average Orders',
 'Fewest Orders',
 'Fewest Orders',
 'Fewest Orders',
 'Fewest Orders',
 'Most Orders',
 'Most Orders',
 'Average Orders',
 'Average Orders',
 'Average Orders',
 'Average Orders',
 'Average Orders',
 'Average Orders',
 'Average Orders',
 'Fewest Orders',
 'Average Orders',
 'Most Orders',
 'Most Orders',
 'Average Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Average Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Average Orders',
 'Most Orders',
 'Average Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Most Orders',
 'Average Orders',
 'Average Orders',
 'Most Orders',
 'Most Ord

In [50]:
#Create "Busiest Period of Day" Column
df_ords_prods_merge['busiest_period_of_day'] = result

In [51]:
#Check Outputs
df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most Orders       21118071
Average Orders     9997651
Fewest Orders      1289137
Name: count, dtype: int64

The grouped sums match the individual times of day added together showing that the loop was operated correctly. Also, there are no rows that have missing data also showing it was done correctly. 

### Export File

In [59]:
#Export ords_prods_merge file
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_new.pkl'))